trivial-builders: rework runCommand' into runCommandWith && expose

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann commented:
Should probably remove `runCommandInStdenvLocal` due to a similar rationale as `runCommandCCLocal`.

Profpatsch reviewed (changes requested):
I’d say this is the point where we should introduce a generic version that takes an attrset of these options, instead of creating a matrix of everything.

`runCommandWith { stdenv = … }` …

Profpatsch commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -42,13 +27,56 @@ rec {
+      name = lib.strings.sanitizeDerivationName name;
+      inherit buildCommand;
+      passAsFile = [ "buildCommand" ]
+        ++ (derivationArgs.passAsFile or []);
```
I’d do this change in a second commit.

Profpatsch replied:
That way the first commit is a pure refactor for the existing commands, and the second changes behavior.

Profpatsch reviewed (changes requested):
Looks good!

sternenseemann committed:
trivial-builders: merge passAsFile with env in runCommand'

Previously it was awkward to use the runCommand-variants with
passAsFile as a double definition of passAsFile would potentially
break runCommand: passAsFile would overwrite the previous definition,
defeating the purpose of setting it in runCommand in the first place.
This is now fixed by concatenating the [ "buildCommand" ] list with
one the one from env, if present.

Adjust buildEnv where passAsFile = null; was passed in some cases,
breaking evaluation since it'd evaluate to [ "buildCommand" ] ++ null.

sternenseemann commented:
New stuff that works:

```nix
with (import ./. {});

runCommandWith {
  # llvm stdenv rebuild on master rn
  stdenv = (import <nixpkgs> {}).llvmPackages.stdenv;
  name = "lol";
} "echo $CC > $out"
# $out will be "clang"
```

```nix
with (import ./. {});

runCommandWith {
  name = "lol";
  derivationArgs = {
    passAsFile = [ "foo" ];
    foo = "bar";
  };
} "mv $fooPath $out"
# $out will be "bar"
```

Profpatsch commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -43,13 +27,60 @@ rec {
-  runCommandCC = runCommand' false stdenv;
+  runCommandCC = name: env: runCommandWith {
+    runLocal = false;
+    inherit stdenv;
```
It feels a bit hard to follow to me to inherit from the file scope, I’d say `stdenv = stdenv` here is okay.

sternenseemann replied:
It's a shame can't have `pkgs` as an argument to this file, really.

Profpatsch commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -43,13 +27,60 @@ rec {
+      # prevent infinite recursion with default no stdenv given
+      defaultStdenv = stdenv;
+    in
+    { stdenv ? defaultStdenv
+    , runLocal ? false
+    , derivationArgs ? {}
+    , name
```
document these with comments I’d say

Profpatsch reviewed (approved):
two small comments, then LGTM

sternenseemann commented:
@ofborg eval

sternenseemann commented:
/rebase staging

sternenseemann committed:
trivial-builders: rework runCommand' into runCommandWith && expose

runCommandWith receives an attribute set with options which previously
were positional arguments of runCommand' and a buildCommand. This
allows for overriding the used stdenv freely (so stuff like
llvmPackages.stdenv can be used). Additionally the possibility to change
arguments passed to stdenv.mkDerivation is made more explicit via the
derivationArgs argument.

Profpatsch commented:
I’d say we can merge.
